added a bullet dict class as #27 requested

CurtisPreston99 opened the pull request:
the way this works is you define a dict the keys are shown to the user but the values are returned when selected. i think this would be nice to have.
added it to client file as class BulletDict, and set up in init, also added an example feel free to give me advice on how to improve this class

CurtisPreston99 committed:
added a bullet dict class as #27 requested

the way this works is you define a dict the keys are shown to the user but the values are returned when selected. i think this would be nice to have
added it to client file as class BulletDict, and set up in init, also added an example feel free to give me advice on how to improve this class

rcfox commented on bullet/__init__.py:
```diff
@@ -6,4 +6,5 @@
-from .client import ScrollBar
\ No newline at end of file
+from .client import ScrollBar
+from .client import BulletDic
```
Your class is named BulletDict, not BulletDic

rcfox commented on examples/BulletDic.py:
```diff
@@ -0,0 +1,16 @@
+from bullet import BulletDic
```
Class name again

rcfox commented on bullet/client.py:
```diff
@@ -224,6 +224,127 @@ def launch(self, default = None):
 
+
+@keyhandler.init
+class BulletDict:
```
This is a lot of repeated code for what seems like it should just be a child class that overrides accept().

CurtisPreston99 replied:
ah okay that makes sense ill do that now thank you.

CurtisPreston99 replied:
what are your thoughts on it now? any improvements you would recommend me trying to do?

CurtisPreston99 committed:
adding inheritance to BulletDict from Bullet and fixing typos

fixing what was suggested by rcfox on pull request #27

rcfox commented:
(I should point out that I have no authority on this project, I'm just a random guy commenting on someone else's repo.)

After sleeping on this, it seems like it would be best if this were rolled into the main classes. (Sorry for the runaround) It's simple enough to detect the type of variable being passed into the constructor.

It also seems like list choices would be the special case. If you get a list, you could just make a dict out of it: `{c: c for c in choices}` With that, we could standardize the rest of the code on dictionaries.

rcfox commented:
Actually, it would probably be better to store choices as a list of (key, value) pairs. That way duplicate list items won't be lost, and you avoid the weirdness of trying to index into a dictionary at a certain position.

wizord-gaming commented:
thats okay im just trying to improve my code, it was my new years resolution to contribute to things, also im still at University so learning better ways of doing things is always nice. also that two list idea seams pretty good so you would just have a list for displaying and a list for return values

CurtisPreston99 committed:
cleaning up

CurtisPreston99 committed:
added dict capabilities to bullet class

CurtisPreston99 committed:
Merge branch 'addingdictToBullet'

rcfox commented on bullet/client.py:
```diff
@@ -131,16 +131,24 @@ def __init__(
+        
+        #test flag also return objects
+        self.objects=[]
+        if(type(choices) is dict):
```
`isinstance(obj, cls)` is generally preferable to `type(obj) is cls` since it supports subtypes. (For instance, `OrderedDict`)

rcfox commented on bullet/client.py:
```diff
@@ -131,16 +131,24 @@ def __init__(
+            self.objects=[choices[x] for x in self.choices]
+        else:
+            self.choices = choices
+        
```
If you set `self.objects = choices`, you wouldn't need the extra check in accept(), you could also just return from self.objects

rcfox commented on bullet/client.py:
```diff
@@ -131,16 +131,24 @@ def __init__(
+        self.objects=[]
+        if(type(choices) is dict):
+            self.choices = list(choices.keys())
+            self.objects=[choices[x] for x in self.choices]
```
How about `list(choices.values())` for consistency?

CurtisPreston99 replied:
did not know that there was a .values() function

rcfox commented on bullet/client.py:
```diff
@@ -197,7 +205,11 @@ def moveDown(self):
         cursor.show_cursor()
-        ret = self.choices[self.pos]
+        #flag check and return ether objects or just choices
+        if(self.objects):
```
Nitpicking, but you generally leave off parentheses on if statements in Python for simple conditions like this.

Also, try to be consistent with whitespace to match the surrounding code.

rcfox reviewed (commented):
Mostly looks good, just needs a bit of polish.

CurtisPreston99 committed:
just some polishing 

removed flag statement from accept
by making objects the same as choices if its a list

tdh8316 commented:
You should use tools such as [PEP8](https://www.python.org/dev/peps/pep-0008/) to maintain a consistent code style.

I think [pylint](https://www.pylint.org/), [yapf](https://github.com/google/yapf) would be good for this :)

wizord-gaming commented:
> You should use tools such as [PEP8](https://www.python.org/dev/peps/pep-0008/) to maintain a consistent code style.
> 
> I think [pylint](https://www.pylint.org/), [yapf](https://github.com/google/yapf) would be good for this :)

ah okay ill look into this
